apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/spi/springsecurity/ApolloPasswordEncoderFactory.java

```diff
@@ -0,0 +1,76 @@
+
+import com.ctrip.framework.apollo.portal.spi.oidc.PlaceholderPasswordEncoder;
+import java.util.HashMap;
+import java.util.Map;
+import org.springframework.security.crypto.argon2.Argon2PasswordEncoder;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.factory.PasswordEncoderFactories;
+import org.springframework.security.crypto.password.DelegatingPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.crypto.password.Pbkdf2PasswordEncoder;
+import org.springframework.security.crypto.scrypt.SCryptPasswordEncoder;
+
+/**
+ * @author vdisk <vdisk@foxmail.com>
+ */
+public final class ApolloPasswordEncoderFactory {
+
+  private ApolloPasswordEncoderFactory() {
+  }
+
+  /**
+   * Creates a {@link DelegatingPasswordEncoder} with default mappings {@link
+   * PasswordEncoderFactories#createDelegatingPasswordEncoder()}, and add a placeholder encoder for
+   * oidc {@link PlaceholderPasswordEncoder}
+   *
+   * @return the {@link PasswordEncoder} to use
+   */
+  @SuppressWarnings("deprecation")
+  public static PasswordEncoder createDelegatingPasswordEncoder() {
+    // copy from PasswordEncoderFactories, and it's should follow the upgrade of the PasswordEncoderFactories
+    String encodingId = "bcrypt";
+    Map<String, PasswordEncoder> encoders = new HashMap<>();
+    encoders.put(encodingId, new BCryptPasswordEncoder());
+    encoders.put("ldap", new org.springframework.security.crypto.password.LdapShaPasswordEncoder());
+    encoders.put("MD4", new org.springframework.security.crypto.password.Md4PasswordEncoder());
+    encoders.put("MD5",
+        new org.springframework.security.crypto.password.MessageDigestPasswordEncoder("MD5"));
+    encoders.put("noop",
+        org.springframework.security.crypto.password.NoOpPasswordEncoder.getInstance());
+    encoders.put("pbkdf2", new Pbkdf2PasswordEncoder());
+    encoders.put("scrypt", new SCryptPasswordEncoder());
+    encoders.put("SHA-1",
+        new org.springframework.security.crypto.password.MessageDigestPasswordEncoder("SHA-1"));
+    encoders.put("SHA-256",
+        new org.springframework.security.crypto.password.MessageDigestPasswordEncoder("SHA-256"));
+    encoders
+        .put("sha256", new org.springframework.security.crypto.password.StandardPasswordEncoder());
+    encoders.put("argon2", new Argon2PasswordEncoder());
+
+    // placeholder encoder for oidc
+    encoders.put(PlaceholderPasswordEncoder.ENCODING_ID, new PlaceholderPasswordEncoder());
+    DelegatingPasswordEncoder delegatingPasswordEncoder = new DelegatingPasswordEncoder(encodingId,
+        encoders);
+
+    // todo: adapt the old password, and it should be removed in the next feature version of the 1.9.x
+    delegatingPasswordEncoder.setDefaultPasswordEncoderForMatches(new PasswordEncoderAdapter(encoders.get(encodingId)));
```

Comment:
Does it work if there is no PasswordEncoderAdapter?

```java
    delegatingPasswordEncoder.setDefaultPasswordEncoderForMatches(encoders.get(encodingId));
```

Reply:
> Does it work if there is no PasswordEncoderAdapter?
> 
> ```java
> delegatingPasswordEncoder.setDefaultPasswordEncoderForMatches(encoders.get(encodingId));
> ```

it do works, but the `BCryptPasswordEncoder` can not throw an exact exception while the password has a wrong prefix just like `{foolish}xxx`. the `PasswordEncoderAdapter` is just used to detect the wrong prefix. see the `org.springframework.security.crypto.password.DelegatingPasswordEncoder.UnmappedIdPasswordEncoder`